### PR TITLE
support intersect filters on raster columns (close #2613)

### LINKED ISSUE #2613: Support Raster Queries
This is a **Feature Request**. Raster support seems to be lacking from all postgis functions supporting it. They are a critical part of many geospatial services. Most needed is the ability to retrieve raster data with `st_intersects`. In particular, with the signature `ST_Intersects( geometry geommin , raster rast , integer nband=NULL )`

This functionality would support queries like this:
```
query getIntersectingField($input: geometry) {
  2d_field(where: {rast: {_st_intersects: {from: $input, args: ...}}}) {
    rid
    rast
}
}
```

### PR DESCRIPTION
**Pull request by rakeshkky**
### Description

This PR has support for raster `ST_Intersects` mentioned [here](https://github.com/hasura/graphql-engine/issues/2613#issuecomment-518126338)

Create the following Postgres schema:
```sql
CREATE TABLE dummy_raster (
  id serial primary key,
  rast_col raster
)
```

The GraphQL Engine generates the following operators for `raster` columns in boolean expressions.

1. `_st_intersects_rast` -> `boolean ST_Intersects(raster <raster-col>, raster <raster-input>)`

```graphql
query ($raster_input: raster) {
  dummy_raster(where: {rast_col: {_st_intersects_rast: $raster_input}}){
    id
    rast_col
  }
}
```
Note:- `raster_input` is a raw `String` value of raster data.

2. `_st_intersects_nband_geom` -> `boolean ST_Intersects(raster <raster-col>, integer nband, geometry geommin)`

```graphql
query ($geom: geometry!) {
  dummy_raster(where: {rast_col: {_st_intersects_nband_geom: {nband: 2, geommin: $geom}}}){
    id
    rast_col
  }
}
```

3. `_st_intersects_geom_nband` -> `boolean ST_Intersects(raster <raster-col> , geometry geommin, integer nband=NULL)`

With `nband` value:
```graphql
query ($geom: geometry!) {
  dummy_raster(where: {rast_col: {_st_intersects_geom_nband: { geommin: $geom, nband: 3}}}){
    id
    rast_col
  }
}
```
Without `nband` value:
```graphql
query ($geom: geometry!) {
  dummy_raster(where: {rast_col: {_st_intersects_geom_nband: { geommin: $geom}}}){
    id
    rast_col
  }
}
```

I'm open to support any other operators/functions related to raster columns. Please comment down below specifying your requirement. 

### Affected components 

- Server
- Docs
- Tests

### Related Issues

close #2613 

### Solution and Design

- Add `PGRaster` constructor to `PGColType` type to represent Postgres `raster` type. Similarly `PGValRaster` for `PGColValue` type
- Generate and resolve `_st_intersects_rast`, `_st_intersects_nband_geom` and `_st_intersects_geom_nband` operators in boolean expressions. 

### Steps to test and verify

Follow any test case added in this pull.

### CODE REVIEW

File: server/src-lib/Hasura/SQL/Value.hs [hunk 1]
```diff
@@ -43,6 +43,7 @@ data PGColValue
   | PGValJSON !Q.JSON
   | PGValJSONB !Q.JSONB
   | PGValGeo !GeometryWithCRS
+  | PGValRaster !T.Text
```
Comment (lexi-lambda) [inline server/src-lib/Hasura/SQL/Value.hs]: I think it would be nice to use a more specific type than `Text` here. Looking at the PostGIS documentation, the format of raster literals appears to use a hexadecimal representation of the [WellKnownBinary (WKB) format](https://github.com/gravitystorm/postgis/blob/master/raster/doc/RFC2-WellKnownBinaryFormat). While it probably doesn’t make sense to actually parse the binary format itself, maybe it would be nice to parse the hex string, since that’s easy to do. Using the [`text-conversions`](https://hackage.haskell.org/package/text-conversions-0.3.0/docs/Data-Text-Conversions.html) package (disclaimer: it’s my package!), we could define

```hs
newtype RasterWKB = RasterWKB { getRasterWKB :: Base16 ByteString }
```

and even derive `FromText` and `ToText` instances for free. That would give us at least slightly more type safety and possibly also slightly better error reporting if users supply invalid raster literals.

File: server/src-lib/Hasura/SQL/Types.hs [hunk 2]
```diff
@@ -378,6 +382,7 @@ pgTypeOid PGJSONB       = PTI.jsonb
 -- we are using the ST_GeomFromGeoJSON($i) instead of $i
 pgTypeOid PGGeometry    = PTI.text
 pgTypeOid PGGeography   = PTI.text
+pgTypeOid PGRaster      = PTI.auto
```
Comment (lexi-lambda) [inline server/src-lib/Hasura/SQL/Types.hs]: Should we use [`RT_ST_RastFromHexWKB`](https://postgis.net/docs/RT_ST_RastFromHexWKB.html) or [`RT_ST_RastFromWKB`](https://postgis.net/docs/RT_ST_RastFromWKB.html) for raster literals, the same way we use `ST_GeomFromGeoJSON` for geometry types?

File: docs/graphql/manual/queries/query-filters.rst [hunk 3]
```diff
@@ -1508,3 +1508,157 @@ Columns of type ``geography`` are more accurate, but they don’t support as man
   .. code-block:: sql
 
     CREATE INDEX cities_location_geography ON cities USING GIST ((location::geography));
+
+Intersect operators on RASTER columns
+-------------------------------------
+
+Intersect operators on columns with ``raster`` type are supported.
+Refer `Postgis docs <https://postgis.net/docs/RT_ST_Intersects.html>`__ to know more about intersect functions on ``raster`` columns.
```
Comment (marionschleifer) [inline docs/graphql/manual/queries/query-filters.rst]: Refer *to*...

File: docs/graphql/manual/queries/query-filters.rst [hunk 4]
```diff
@@ -1508,3 +1508,157 @@ Columns of type ``geography`` are more accurate, but they don’t support as man
...
+Please submit feature request via `github <https://github.com/hasura/graphql-engine>`__ if you want support for more functions.
```
Comment (marionschleifer) [inline docs/graphql/manual/queries/query-filters.rst]: Please submit *a* feature request...

File: docs/graphql/manual/queries/query-filters.rst [hunk 5]
```diff
@@ -1508,3 +1508,157 @@ Columns of type ``geography`` are more accurate, but they don’t support as man
...
+
+Example: _st_intersects_rast
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Filter the raster values which intersects the input raster value.
```
Comment (marionschleifer) [inline docs/graphql/manual/queries/query-filters.rst]: Filter the raster values which *intersect*...

File: docs/graphql/manual/queries/query-filters.rst [hunk 6]
```diff
@@ -1508,3 +1508,157 @@ Columns of type ``geography`` are more accurate, but they don’t support as man
...
+           "name": "urn:ogc:def:crs:EPSG::4326"
+         }
+       }
+     }
+   }
+
+Example: _st_intersects_nband_geom
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Filter the raster values which intersects the input geometry value and with provided band number.
```
Comment (marionschleifer) [inline docs/graphql/manual/queries/query-filters.rst]: Filter the raster values which *intersect* the...
I don't understand the end of the sentence i.e. what happens with the band number? Could you rephrase that?

File: server/src-lib/Hasura/SQL/Value.hs (hunk 1 above)
Comment (rakeshkky) [reply server/src-lib/Hasura/SQL/Value.hs]: Thank you. This helps me a lot. :slightly_smiling_face:
